src/Compilers/CSharp/Test/Symbol/Symbols/CovariantReturnTests.cs

```diff
@@ -1659,11 +2115,793 @@ End Class
+  IL_0101:  call       ""Function System.Reflection.MethodBase.GetMethodFromHandle(System.RuntimeMethodHandle) As System.Reflection.MethodBase""
+  IL_0106:  castclass  ""System.Reflection.MethodInfo""
+  IL_010b:  ldc.i4.2
+  IL_010c:  newarr     ""System.Linq.Expressions.Expression""
+  IL_0111:  dup
+  IL_0112:  ldc.i4.0
+  IL_0113:  ldtoken    ""System.Func(Of String)""
+  IL_0118:  call       ""Function System.Type.GetTypeFromHandle(System.RuntimeTypeHandle) As System.Type""
+  IL_011d:  ldtoken    ""System.Type""
+  IL_0122:  call       ""Function System.Type.GetTypeFromHandle(System.RuntimeTypeHandle) As System.Type""
+  IL_0127:  call       ""Function System.Linq.Expressions.Expression.Constant(Object, System.Type) As System.Linq.Expressions.ConstantExpression""
+  IL_012c:  stelem.ref
+  IL_012d:  dup
+  IL_012e:  ldc.i4.1
+  IL_012f:  ldloc.s    V_15
+  IL_0131:  ldtoken    ""Object""
+  IL_0136:  call       ""Function System.Type.GetTypeFromHandle(System.RuntimeTypeHandle) As System.Type""
+  IL_013b:  call       ""Function System.Linq.Expressions.Expression.Convert(System.Linq.Expressions.Expression, System.Type) As System.Linq.Expressions.UnaryExpression""
+  IL_0140:  stelem.ref
+  IL_0141:  call       ""Function System.Linq.Expressions.Expression.Call(System.Linq.Expressions.Expression, System.Reflection.MethodInfo, ParamArray System.Linq.Expressions.Expression()) As System.Linq.Expressions.MethodCallExpression""
+  IL_0146:  ldtoken    ""System.Func(Of String)""
+  IL_014b:  call       ""Function System.Type.GetTypeFromHandle(System.RuntimeTypeHandle) As System.Type""
+  IL_0150:  call       ""Function System.Linq.Expressions.Expression.Convert(System.Linq.Expressions.Expression, System.Type) As System.Linq.Expressions.UnaryExpression""
+  IL_0155:  ldc.i4.1
+  IL_0156:  newarr     ""System.Linq.Expressions.ParameterExpression""
+  IL_015b:  dup
+  IL_015c:  ldc.i4.0
+  IL_015d:  ldloc.s    V_15
+  IL_015f:  stelem.ref
+  IL_0160:  call       ""Function System.Linq.Expressions.Expression.Lambda(Of System.Func(Of Derived, System.Func(Of String)))(System.Linq.Expressions.Expression, ParamArray System.Linq.Expressions.ParameterExpression()) As System.Linq.Expressions.Expression(Of System.Func(Of Derived, System.Func(Of String)))""
+  IL_0165:  stloc.s    V_14
+  // sequence point: End Sub
+  IL_0167:  ret
+}
+");
+        }
+
+        [Fact]
+        public void BinaryCompatibility_01()
+        {
+            var s0 = @"
+public class Base
+{
+    public virtual object M() => null;
+}
+";
+            var ref0 = CreateCompilationWithoutCovariantReturns(s0).EmitToImageReference();
+
+            var s1a = @"
+public class Mid : Base
+{
+}
+";
+            var ref1a = CreateCompilationWithoutCovariantReturns(s1a, references: new[] { ref0 }, assemblyName: "ref1").EmitToImageReference();
+
+            var s1b = @"
+public class Mid : Base
+{
+    public override string M() => null;
```

Comment:
>public override string M() => null; [](start = 4, length = 35)

I think we should also test scenario when this is:
- ```new virtual method```, I assume it will be newSlot
- ```new``` but not virtual, I assume it will be newSlot
- all of the above including original but when there is no hidebysig flag in metadata. 

Also when the method overrides explicitly, but not the Base.M, but rather a method higher in the type hierarchy.

Reply:
The "Also when the method overrides explicitly, but not the Base.M, but rather a method higher in the type hierarchy." is what is tested on `Derived.M` in `verify2` below.

I have added the other suggestions to the test plan.

---
In reply to: [424780749](https://github.com/dotnet/roslyn/pull/44025#discussion_r424780749) [](ancestors = 424780749)